File: multiplication.py
MOD = 0x3b800001
W = [31]
for _ in range(22): W.append( W[-1]**2%MOD )
def dft(poly,b):
    n = 1<<b
    v = poly + [0]*(n-len(poly))
    for loop in range(1,b+1):
        nv = [0]*n
        acc = 1
        space = n>>loop
        for q in range(0,n,space):
            for r in range(space):
                nv[q+r] = (v[2*q%n + r] + acc*v[2*q%n + space + r])%MOD
            acc = acc*W[-loop]%MOD
        v = nv
    return v
def mul(a,b):
    j = (len(a)+len(b)-2).bit_length()
    A,B = dft(a,j), dft(b,j)
    inv = pow(1<<j,-1,MOD)
    C = [A[i]*B[i]*inv%MOD for i in range(1<<j)]
    D = dft(C,j)
    return D[:1] + D[:-len(a)-len(b)+1:-1]

"""
N,M = [int(t) for t in input().split()]
a = [int(t) for t in input().split()]
b = [int(t) for t in input().split()]
c = mul(a,b)
print(*c)
"""

File: cpp/multiplication.cpp
const array<ll,23> W = []{
	array<ll,23> W = {2200}; //order 2^23
	rep(i,0,22) W[i+1] = W[i]*W[i]%MOD;
	return W;
}();
vector<ll> dft( vector<ll> poly, int b ){
	int n = 1<<b;
	poly.resize(n,0);

	rep(loop,1,b+1){
		vector<ll> nv(n,0);

		ll acc = 1;
		int space = n>>loop;
		for( int q = 0; q < n; q += space ){
			rep(r,0,space) nv[q+r] = (poly[2*q%n + r] + acc*poly[2*q%n + space + r])%MOD;
			acc = acc*W[23-loop]%MOD;
		}
		poly = nv;
	}
	return poly;
}
vector<ll> mul( vector<ll> a, vector<ll> b ){
	int sz = a.size() + b.size() - 1;
	int j = 1;
	while( 1<<j < sz ) j++;
	
	vector<ll> A = dft(a,j);
	vector<ll> B = dft(b,j);
	ll inv = p0w(1<<j,MOD-2);
	rep(i,0,B.size()) B[i] = A[i]*B[i]%MOD*inv%MOD;
	
	A = dft(B,j);
	reverse(A.begin()+1,A.end());
	A.resize(sz);
	return A;
}
